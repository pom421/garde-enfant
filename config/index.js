export const CAPACITY_NORMAL_HOURS = 40
export const CAPACITY_EXTRA_HOURS = 8

export const NB_HOURS_FOR_VACATION = 8
export const HOURS_REGULAR_DAY = 8

export const REASONS_ABSENCE = {
  WE: "WE",
  VACATION: "VACATION",
  UNPAID_VACATION: "UNPAID_VACATION",
  RTT: "RTT",
  PUBLIC_HOLIDAY: "PUBLIC_HOLIDAY",
  SICKNESS: "SICKNESS",
  MATERNITY: "MATERNITY",
  OTHER: "OTHER",
}
