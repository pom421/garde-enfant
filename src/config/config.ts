import type { REASONS_ABSENCE } from "@/types/types"

export const CAPACITY_NORMAL_HOURS = 40
export const CAPACITY_EXTRA_HOURS_25 = 8

export const INFOS_ABSENCE: Record<REASONS_ABSENCE, { label: string; nbHours: number }> = {
  WE: { label: "WE", nbHours: 0 },
  VACATION: { label: "Congé payé", nbHours: 0 },
  UNPAID_VACATION: { label: "Congé sans solde", nbHours: 0 },
  PAID_VACATION: { label: "Congé payé exceptionnel", nbHours: 0 }, // ex: évènements familiaux, décès d'un proche, absence pour convenances personnelles des parents.
  PUBLIC_HOLIDAY: { label: "Jour férié", nbHours: 0 }, // (+10% pour les jours fériés et travaillés)
  SICKNESS: { label: "Arrêt maladie", nbHours: 0 }, // La salariée reçoit des indemnités de la CPAM.
  MATERNITY: { label: "Congé maternité", nbHours: 0 },
}

export const PAID_ABSENCES = ["VACATION", "PAID_VACATION", "PUBLIC_HOLIDAY"]
export const UNPAID_ABSENCES = ["UNPAID_VACATION", "SICKNESS", "MATERNITY"]

export function colorAbsence(reasonAbsence: REASONS_ABSENCE) {
  switch (reasonAbsence) {
    case "WE":
      return "gray.100"
    case "VACATION":
      return "teal.100"
    case "SICKNESS":
      return "orange.100"
    case "PUBLIC_HOLIDAY":
      return "blue.100"
    default:
      return "white"
  }
}
