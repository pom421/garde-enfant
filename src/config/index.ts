export const CAPACITY_NORMAL_HOURS = 40
export const CAPACITY_EXTRA_HOURS_25 = 8

export const NB_HOURS_FOR_VACATION = 8
export const HOURS_REGULAR_DAY = 10

export type REASONS_ABSENCE =
  | "WE"
  | "VACATION"
  | "UNPAID_VACATION"
  | "RTT"
  | "PUBLIC_HOLIDAY"
  | "SICKNESS"
  | "MATERNITY"
  | "OTHER"

export const LABEL_ABSENCE: Record<REASONS_ABSENCE, string> = {
  WE: "WE",
  VACATION: "Congé payé",
  UNPAID_VACATION: "Congé sans solde",
  RTT: "RTT",
  PUBLIC_HOLIDAY: "Jour férié",
  SICKNESS: "Arrêt maladie",
  MATERNITY: "Congé maternité",
  OTHER: "Autre",
}

export function colorAbsence(reasonAbsence: REASONS_ABSENCE) {
  switch (reasonAbsence) {
    case "WE":
      return "gray.100"
    case "VACATION":
      return "teal.100"
    case "SICKNESS":
      return "orange.100"
    default:
      return "white"
  }
}
