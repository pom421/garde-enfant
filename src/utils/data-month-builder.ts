import { isWeekend, isWithinInterval, parseISO } from "date-fns"

import type { AbsenceType, DataWeeksType, REASONS_ABSENCE } from "@/types/types"

import { INFOS_ABSENCE } from "@/config/config"
import { hoursRegularDay } from "@/data/app"
import { assertDate, YearMonthType } from "@/utils/date"
import { helpersForMonth } from "@/utils/month"
import { isPublicHoliday } from "@/utils/public-holidays"

// Construction de la structure de données pour le mois : tableau représentant les semaines, chaque semaine contenant un tableau de jours.
export function buildDataWeeks({ absences = [] } = { absences: [] }) {
  return function (yearMonth: YearMonthType): DataWeeksType {
    const { weeks } = helpersForMonth(yearMonth)

    const fillDay = fillDayBuilder({ absences })

    return weeks.map((week) => ({
      days: week.map((date) => fillDay(date)),
    }))
  }
}

/**
 * Fonction qui remplit les données du jour.
 *
 * @param {Object} options. hours, la liste des heures modifiées de la garde d'enfant. absences, la liste des absences de la garde d'enfant.
 */
export function fillDayBuilder({ absences = [] }: { absences?: AbsenceType[] }) {
  return function (date): { date: Date; reasonAbsence?: REASONS_ABSENCE; nbHours: number } {
    assertDate(date)

    // la date est-elle un samedi ou dimanche?
    if (isWeekend(date))
      return {
        date,
        nbHours: 0,
        reasonAbsence: "WE",
      }
    // la date est-elle un jour férié ?
    if (isPublicHoliday(date))
      return {
        date,
        nbHours: 0,
        reasonAbsence: "PUBLIC_HOLIDAY",
      }

    // la date est-elle une absence ?
    for (const absence of absences) {
      if (isWithinInterval(date, { start: parseISO(absence.start), end: parseISO(absence.end) })) {
        return {
          date,
          nbHours: INFOS_ABSENCE[absence.reason].nbHours,
          reasonAbsence: absence.reason,
        }
      }
    }

    // par défaut, on ajoute le nombre d'heure standard.
    return {
      date,
      nbHours: hoursRegularDay,
    }
  }
}
