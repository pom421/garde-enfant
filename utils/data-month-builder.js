import { isWeekend, isWithinInterval, isSameDay, parseISO } from "date-fns"
import { formatInTimeZone, utcToZonedTime } from "date-fns-tz"

import { utilMonth } from "@/utils/month"
import { isPublicHoliday } from "@/utils/public-holidays"
import { CAPACITY_NORMAL_HOURS, CAPACITY_EXTRA_HOURS, HOURS_REGULAR_DAY } from "@/config/index"
import { REASONS_ABSENCE } from "@/config/index"
import { assertDate } from "./date"

function displayWeek(numWeek, days) {
  // eslint-disable-next-line no-console
  console.log("week", numWeek)

  for (const day of days) {
    // eslint-disable-next-line no-console
    console.log(formatInTimeZone(day, "Europe/Paris", "yyyy-MM-dd"))
  }
}

// Construction de la structure de données pour le mois.
export function buildDataMonth({ hours, absences }) {
  return function ({ year, month }) {
    const dataMonth = {
      meta: {
        year,
        month,
        normalHours: 0,
        extraHours: 0,
        moreExtraHours: 0,
      },
      lastWeekPreviousMonth: {},
      weeks: [],
    }

    const { lastWeekPreviousMonth, weeks } = utilMonth({ year, month })

    dataMonth.lastWeekPreviousMonth = {
      days: lastWeekPreviousMonth.map((day) => ({
        date: day,
        nbHour: 0,
        reasonAbsence: "",
      })),
      hours: {
        capacityNormalHours: CAPACITY_NORMAL_HOURS,
        capacityExtraHours: CAPACITY_EXTRA_HOURS,
        totalHours: 0,
        normalHours: 0,
        extraHours: 0,
        moreExtraHours: 0,
      },
    }

    dataMonth.weeks = weeks.map((week) => ({
      days: week.map((day) => ({
        date: day,
        nbHours: 0,
        reasonAbsence: "",
      })),
      hours: {
        capacityNormalHours: CAPACITY_NORMAL_HOURS,
        capacityExtraHours: CAPACITY_EXTRA_HOURS,
        totalHours: 0,
        normalHours: 0,
        extraHours: 0,
        moreExtraHours: 0,
      },
    }))

    dataMonth.lastWeekPreviousMonth.days = dataMonth.lastWeekPreviousMonth.days.map((day) =>
      fillDay({ hours, absences })(day.date),
    )

    dataMonth.weeks.forEach((week) => {
      week.days = week.days.map((day) => fillDay({ hours, absences })(day.date))
    })

    return dataMonth
  }
}

/**
 * Fonction qui remplit les données du jour.
 *
 * @param {Object} options. hours, la liste des heures modifiées de la garde d'enfant. absences, la liste des absences de la garde d'enfant.
 */
export function fillDay({ hours = [], absences = [] }) {
  return function (date) {
    assertDate(date)

    // la date est-elle un samedi ou dimanche?
    if (isWeekend(date))
      return {
        date,
        reasonAbsence: REASONS_ABSENCE.WE,
      }
    // la date est-elle un jour férié ?
    if (isPublicHoliday(date))
      return {
        date,
        reasonAbsence: REASONS_ABSENCE.PUBLIC_HOLIDAY,
      }

    // la date a-t-elle des heures explicites ?
    const [entryHours] = hours.filter((hour) => isSameDay(parseISO(hour.date), date)) // on considère qu'il ne peut pas y avoir 2 entrées pour la même date.

    if (entryHours) {
      return {
        date,
        nbHours: entryHours.nbHours,
      }
    }
    // la date est-elle une absence ?
    // eslint-disable-next-line prettier/prettier
    for (const absence of absences) {
      if (isWithinInterval(date, { start: parseISO(absence.start), end: parseISO(absence.end) })) {
        return {
          date,
          reasonAbsence: absence.reason,
        }
      }
    }

    // par défaut, on ajoute le nombre d'heure standard.
    return {
      date,
      nbHours: HOURS_REGULAR_DAY,
    }
  }
}

/*
const dataMonth = {
  normalHours: 200,
  extraHours: 32,
  moreExtraHours: 10,
  lastWeekPreviousMonth: {
    days: [
      {
        date: "2022-01-31",
        reasonAbsence: "",
        nbHours: 10,
      },
    ],
    capacityNormalHours: 40,
    capacityExtraHours: 8,
    normalHours: 10,
    extraHours: 0,
    moreExtraHours: 0,
  },
  weeks: [
    {
      days: [
        {
          date: "2022-02-01",
          nbHours: 10,
        },
        {
          date: "2022-02-02",
          nbHours: 10,
        },
        {
          date: "2022-02-03",
          nbHours: 10,
        },
        {
          date: "2022-02-04",
          nbHours: 10,
        },
        {
          date: "2022-02-05",
          reasonAbsence: REASONS_ABSENCE.WE,
        },
        {
          date: "2022-02-06",
          reasonAbsence: REASONS_ABSENCE.WE,
        },
      ],
      capacityNormalHours: 30, // 40 - normalHours de la semaine équivalente du mois précédent
      capacityExtraHours: 8, // 8 - extraHours de la semaine équivalente du mois précédent
      totalHours: 40,
      normalHours: 30,
      extraHours: 8,
      moreExtraHours: 2,
    },
    {
      name: "S2",
      days: [
        {
          date: "2022-02-01",
          reasonAbsence: "",
          nbHours: 10,
        },
        {
          date: "2022-02-02",
          reasonAbsence: "",
          nbHours: 10,
        },
        {
          date: "2022-02-03",
          reasonAbsence: "",
          nbHours: 10,
        },
        {
          date: "2022-02-04",
          reasonAbsence: "",
          nbHours: 10,
        },
        {
          date: "2022-02-05",
          reasonAbsence: REASONS_ABSENCE.WE,
        },
        {
          date: "2022-02-06",
          reasonAbsence: REASONS_ABSENCE.WE,
        },
      ],
      capacityNormalHours: 0, // 40 - normalHours de la semaine équivalente du mois précédent
      capacityExtraHours: 8, // 8 - extraHours de la semaine équivalente du mois précédent
      totalHours: 50,
      normalHours: 30,
      extraHours: 8,
      moreExtraHours: 2,
    },
  ],
}
*/
