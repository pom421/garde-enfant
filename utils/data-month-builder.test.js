import { HOURS_REGULAR_DAY, REASONS_ABSENCE } from "@/config/index"
import { parseISO } from "date-fns"
import { buildDataMonth, fillDay } from "./data-month-builder"

test("should find a sunday", () => {
  const { reasonAbsence, nbHours } = fillDay({})(parseISO("2022-01-30")) // dimanche

  expect(nbHours).toBe(undefined)
  expect(reasonAbsence).toBe(REASONS_ABSENCE.WE)
})

test("should find a vacation", () => {
  const absences = [
    {
      start: "2022-01-30",
      end: "2022-02-01",
      reason: REASONS_ABSENCE.VACATION,
    },
  ]

  const { reasonAbsence, nbHours } = fillDay({ absences })(parseISO("2022-01-31"))

  expect(nbHours).toBe(undefined)
  expect(reasonAbsence).toBe(REASONS_ABSENCE.VACATION)
})

test("should find a publicHoliday", () => {
  const { reasonAbsence, nbHours } = fillDay({})(parseISO("2022-07-14")) // jours férié

  expect(nbHours).toBe(undefined)
  expect(reasonAbsence).toBe(REASONS_ABSENCE.PUBLIC_HOLIDAY)
})

test("should find a regulard day with defaut hours", () => {
  const { reasonAbsence, nbHours } = fillDay({})(parseISO("2022-07-15"))

  expect(nbHours).toBe(HOURS_REGULAR_DAY)
  expect(reasonAbsence).toBe(undefined)
})

// build data month tests ---------------------------- //

test("should match snapshot", () => {
  const hours = [
    {
      date: "2022-01-31",
      nbHours: 5,
    },
    {
      date: "2022-02-10",
      nbHours: 7,
    },
  ]
  const absences = [
    {
      start: "2022-01-30",
      end: "2022-02-05",
      reason: REASONS_ABSENCE.VACATION,
    },
  ]

  const dataMonth = buildDataMonth({ hours, absences })({ year: 2022, month: 1 })

  expect(dataMonth).toMatchSnapshot()
})
