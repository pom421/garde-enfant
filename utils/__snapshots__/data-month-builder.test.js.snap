// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
Object {
  "lastWeekPreviousMonth": Object {
    "days": Array [
      Object {
        "date": 2022-01-30T23:00:00.000Z,
        "nbHours": 5,
      },
    ],
    "hours": Object {
      "capacityExtraHours": 8,
      "capacityNormalHours": 40,
      "extraHours": 0,
      "moreExtraHours": 0,
      "normalHours": 0,
      "totalHours": 0,
    },
  },
  "meta": Object {
    "extraHours": 0,
    "month": 1,
    "moreExtraHours": 0,
    "normalHours": 0,
    "year": 2022,
  },
  "weeks": Array [
    Object {
      "days": Array [
        Object {
          "date": 2022-01-31T23:00:00.000Z,
          "reasonAbsence": "VACATION",
        },
        Object {
          "date": 2022-02-01T23:00:00.000Z,
          "reasonAbsence": "VACATION",
        },
        Object {
          "date": 2022-02-02T23:00:00.000Z,
          "reasonAbsence": "VACATION",
        },
        Object {
          "date": 2022-02-03T23:00:00.000Z,
          "reasonAbsence": "VACATION",
        },
        Object {
          "date": 2022-02-04T23:00:00.000Z,
          "reasonAbsence": "WE",
        },
        Object {
          "date": 2022-02-05T23:00:00.000Z,
          "reasonAbsence": "WE",
        },
      ],
      "hours": Object {
        "capacityExtraHours": 8,
        "capacityNormalHours": 40,
        "extraHours": 0,
        "moreExtraHours": 0,
        "normalHours": 0,
        "totalHours": 0,
      },
    },
    Object {
      "days": Array [
        Object {
          "date": 2022-02-06T23:00:00.000Z,
          "nbHours": 8,
        },
        Object {
          "date": 2022-02-07T23:00:00.000Z,
          "nbHours": 8,
        },
        Object {
          "date": 2022-02-08T23:00:00.000Z,
          "nbHours": 8,
        },
        Object {
          "date": 2022-02-09T23:00:00.000Z,
          "nbHours": 7,
        },
        Object {
          "date": 2022-02-10T23:00:00.000Z,
          "nbHours": 8,
        },
        Object {
          "date": 2022-02-11T23:00:00.000Z,
          "reasonAbsence": "WE",
        },
        Object {
          "date": 2022-02-12T23:00:00.000Z,
          "reasonAbsence": "WE",
        },
      ],
      "hours": Object {
        "capacityExtraHours": 8,
        "capacityNormalHours": 40,
        "extraHours": 0,
        "moreExtraHours": 0,
        "normalHours": 0,
        "totalHours": 0,
      },
    },
    Object {
      "days": Array [
        Object {
          "date": 2022-02-13T23:00:00.000Z,
          "nbHours": 8,
        },
        Object {
          "date": 2022-02-14T23:00:00.000Z,
          "nbHours": 8,
        },
        Object {
          "date": 2022-02-15T23:00:00.000Z,
          "nbHours": 8,
        },
        Object {
          "date": 2022-02-16T23:00:00.000Z,
          "nbHours": 8,
        },
        Object {
          "date": 2022-02-17T23:00:00.000Z,
          "nbHours": 8,
        },
        Object {
          "date": 2022-02-18T23:00:00.000Z,
          "reasonAbsence": "WE",
        },
        Object {
          "date": 2022-02-19T23:00:00.000Z,
          "reasonAbsence": "WE",
        },
      ],
      "hours": Object {
        "capacityExtraHours": 8,
        "capacityNormalHours": 40,
        "extraHours": 0,
        "moreExtraHours": 0,
        "normalHours": 0,
        "totalHours": 0,
      },
    },
    Object {
      "days": Array [
        Object {
          "date": 2022-02-20T23:00:00.000Z,
          "nbHours": 8,
        },
        Object {
          "date": 2022-02-21T23:00:00.000Z,
          "nbHours": 8,
        },
        Object {
          "date": 2022-02-22T23:00:00.000Z,
          "nbHours": 8,
        },
        Object {
          "date": 2022-02-23T23:00:00.000Z,
          "nbHours": 8,
        },
        Object {
          "date": 2022-02-24T23:00:00.000Z,
          "nbHours": 8,
        },
        Object {
          "date": 2022-02-25T23:00:00.000Z,
          "reasonAbsence": "WE",
        },
        Object {
          "date": 2022-02-26T23:00:00.000Z,
          "reasonAbsence": "WE",
        },
      ],
      "hours": Object {
        "capacityExtraHours": 8,
        "capacityNormalHours": 40,
        "extraHours": 0,
        "moreExtraHours": 0,
        "normalHours": 0,
        "totalHours": 0,
      },
    },
    Object {
      "days": Array [
        Object {
          "date": 2022-02-27T23:00:00.000Z,
          "nbHours": 8,
        },
      ],
      "hours": Object {
        "capacityExtraHours": 8,
        "capacityNormalHours": 40,
        "extraHours": 0,
        "moreExtraHours": 0,
        "normalHours": 0,
        "totalHours": 0,
      },
    },
  ],
}
`;
